### PR TITLE
refactor(lib/data): monorepo

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- move preset objects from `monorepo.ts` to `lib/data/monorepo.json`
- add `lib/data/readme`

## Context
Part of: https://github.com/renovatebot/renovate/issues/29942

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository 
https://github.com/Rahul-renovate-testing/extends-monorepo # verifying that monorepo presets work as expected

### CODE REVIEW

File: lib/data/readme.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Introduction
```
Comment (HonkingGoose) [inline lib/data/readme.md:1]: Can you put your latest readme version into this PR? The file is missing a lot of extra improvements. 🙂

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/data/readme.md (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/data/readme.md:1]: Do you mean the information related to other files? Cause apart from that this should be the same.

Comment (RahulGautamSingh) [reply lib/data/readme.md:1]: I have not included the other files as this PR only adds `monorepo.json` file. The others will be added in subsequent PRs.

File: lib/config/presets/internal/monorepos.ts [hunk 2]
```diff
@@ -1,610 +1,35 @@
...
-for (const [name, value] of Object.entries(repoGroups)) {
-  presets[name] = {
-    description: `${name} monorepo`,
-    matchSourceUrls: toArray(value),
-  };
+async function getMonorepoPresets(): Promise<Record<string, Preset>> {
+  const presets: Record<string, Preset> = {};
+  const groups = JSON.parse(
+    await fs.readFile('lib/data/monorepo.json', 'utf8'),
+  );
```
Comment (viceice) [inline lib/config/presets/internal/monorepos.ts:9]: use static jsonc import, so this file is copied automatically. we also don't need async then

**Review by viceice — COMMENTED** [review]
we should create a JSON Schema for it, so it's helping users with tools like vscode

File: lib/workers/global/config/parse/cli.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { Command } from 'commander';
+import pkg from '../../../../../package.json';
```
Comment (viceice) [inline lib/workers/global/config/parse/cli.ts:2]: same here

File: lib/workers/global/index.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import { ERROR } from 'bunyan';
 import fs from 'fs-extra';
 import semver from 'semver';
 import upath from 'upath';
+import pkg from '../../../package.json';
```
Comment (viceice) [inline lib/workers/global/index.ts:6]: same

**Review by viceice — CHANGES_REQUESTED** [review]

File: tsconfig.app.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "extends": "./tsconfig",
   "compilerOptions": {
     "declaration": true,
-    "resolveJsonModule": false,
+    "resolveJsonModule": true,
```
Comment (viceice) [inline tsconfig.app.json:5]: drop this line

File: package.json [hunk 6]
```diff
@@ -4,8 +4,8 @@
   "version": "0.0.0-semantic-release",
   "type": "commonjs",
   "bin": {
-    "renovate": "dist/renovate.js",
-    "renovate-config-validator": "dist/config-validator.js"
+    "renovate": "dist/lib/renovate.js",
+    "renovate-config-validator": "dist/lib/config-validator.js"
```
Comment (viceice) [inline package.json:8]: can't do this, breaking change and unexpected to this PR

**Comment by RahulGautamSingh** [comment]
blocked by: https://github.com/renovatebot/renovate/pull/30223

File: lib/workers/global/index.ts [hunk 7]
```diff
@@ -70,7 +70,7 @@ function haveReachedLimits(): boolean {
 
 /* istanbul ignore next */
 function checkEnv(): void {
-  const range = pkg.engines!.node!;
+  const range = pkg.engines.node;
```
Comment (RahulGautamSingh) [inline lib/workers/global/index.ts:73]:
```suggestion
  const range = pkg.engines!.node!;
```

**Comment by RahulGautamSingh** [comment]
> we should create a JSON Schema for it, so it's helping users with tools like vscode

task for another pr?

**Comment by RahulGautamSingh** [comment]
> we should create a JSON Schema for it, so it's helping users with tools like vscode

task for another pr?

File: lib/data/readme.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+| ------------------------ | --------------- |
+| Source repository URLs   | `repoGroups`    |
+| Organization URls        | `orgGroups`     |
+| Package name patterns(s) | `patternGroups` |
+
+Each method allows you to group related packages based on different criteria:
+
+`repoGroups`: Group packages from the same source repository.
+`orgGroups`: Group packages from the same organization.
+`patternGroups`: Group packages based on name patterns or prefixes.
```
Comment (HonkingGoose) [inline lib/data/readme.md:32]:
```suggestion
| Grouping Criteria        | Method          | Group packages                      |
| ------------------------ | --------------- | ----------------------------------- |
| Source repository URLs   | `repoGroups`    | From the same source repository.    |
| Organization URls        | `orgGroups`     | From the same organization.         |
| Package name patterns(s) | `patternGroups` | Based on name patterns or prefixes. |
```

How about this? It puts the info into the table, and has a short heading for the explanation bit.

**Review by secustor — COMMENTED** [review]
LGTM, outside of docs comment

File: lib/data/readme.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+### Ways to group packages
+
+There are three ways to group packages:
+
+| Method          | Group packages                      |
+| --------------- | ----------------------------------- |
+| `repoGroups`    | From the same source repository.    |
+| `orgGroups`     | From the same organization.         |
+| `patternGroups` | Based on name patterns or prefixes. |
```
Comment (HonkingGoose) [inline lib/data/readme.md:26]: Great idea to put everything into two rows, and removing the duplication! :smile: 

We should put the simple info up-front, like this:

```suggestion
| Group packages                      | Method          |
| ----------------------------------- | --------------- |
| From the same source repository.    | `repoGroups`    |
| From the same organization.         | `orgGroups`     |
| Based on name patterns or prefixes. | `patternGroups` |
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Nearly good to go! :smile:

File: lib/data/readme.md (hunk 8 above)
Comment (HonkingGoose) [reply lib/data/readme.md]: Great idea! I think swapping the order around is even better, see my suggestion below.

You can resolve _this_ comment now though.

**Review by HonkingGoose — APPROVED** [review]
I'm happy with the docs now. :partying_face:

File: tools/docs/presets.ts [hunk 10]
```diff
@@ -9,12 +9,14 @@ function jsUcfirst(string: string): string {
 function getEditUrl(name: string): string {
   const url =
     'https://github.com/renovatebot/renovate/edit/main/lib/config/presets/internal/';
-
+  const newUrl = 'https://github.com/renovatebot/renovate/edit/main/lib/data/';
```
Comment (viceice) [inline tools/docs/presets.ts:12]:
```suggestion
  const dataUrl = 'https://github.com/renovatebot/renovate/edit/main/lib/data/';
```

File: tools/docs/presets.ts [hunk 11]
```diff
@@ -9,12 +9,14 @@ function jsUcfirst(string: string): string {
...
   switch (name) {
     case 'customManagers':
       return `${url}custom-managers.ts`;
     case 'mergeConfidence':
       return `${url}merge-confidence.ts`;
+    case 'monorepo':
+      return `${newUrl}${name}.json`;
```
Comment (viceice) [inline tools/docs/presets.ts:19]:
```suggestion
      return `${dataUrl}${name}.json`;
```

**Review by viceice — COMMENTED** [review]
just a small thing

**Review by viceice — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Pull request merged**